### PR TITLE
[cache] Support async RangeMissingHandler callbacks

### PR DESCRIPTION
**Pull request by arteam**
Change `fillCacheRange` method to accept a completion listener that must be called by `RangeMissingHandler` implementations when they finish fetching data. By doing so, we support asynchronously fetching the data from a third party storage. We also support asynchronous `SourceInputStreamFactory` for reading gaps from the storage.

Resubmition of #110587

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java
Comment (kingherc) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1267]: nit I was wondering whether this should be runAfter or runBefore by comparing with the old code. Probably runBefore makes more sense, so that the region definitely owns IO, else with runAfter it might not own IO due to a race?

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java
Comment (kingherc) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java:76]: nit Isn't this equivalent to runBefore?

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java
Comment (kingherc) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java:1056]: nit String edit is not that readable. Maybe original string is good enough? (similar comment for the other cases where the same string is edited)

**Review by kingherc — COMMENTED** [review]
Looks good. Some nits only. I'll approve after the serverless PR is approved.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java
Comment (arteam) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1267]: For me it's easier to to use `runBefore`, it has the same semantics as a `finally` block, so make sure that it's called inside the context of the handler.

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java
Comment (arteam) [reply x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java:76]: It's equivalent of `runBefore().onResponse(null)` or `ActionRunnable.run(listener, runnable).run()`, but seem more clearer.

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java
Comment (arteam) [reply x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java:1056]: Fixed, thank you! That was an unintended change.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1370]: I wonder if we could skip the change to the `fillCacheRange` method in this PR and focus on just making `SourceInputStreamFactory#create` async. The PR and the associated one in the other repo would become quite a bit simpler. More importantly, for the immediate task at hand, i.e. fetching data from indexing node asynchronously, the blocking concern is only when creating the `InputStream`. Reading from it subsequently in `fillCacheRange` is from an in-memory `BytesReferenceStreamInput` which does not need to be async. For opening and reading input stream from the object store, things can be as is for the time being until s3 client is upgraded to be asynchronous (or maybe when we need to make azure and gcs work). 

I do think we likely need async `fillCacheRange` at some point. My suggestion is that we can asyncify it separately in a future PR which would at least make coding and reviewing the associated PR in the other repo simpler.

Comment (arteam) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1370]: Yeah, in the retrospect it would have been easier just to focus on making `SourceInputStreamFactory` async, but I think it's better to proceed with making `fillCacheRange` async, since we already did a lot of work in making it happen.

**Review by kingherc — APPROVED** [review]
LGTM. Please get a green CI and another approval for this PR and the serverless PR before merging the PRs.

**Comment by kingherc** [comment]
Another review here would be appreciated.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java
Comment (henningandersen) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:1030]: I think I prefer to keep them, since there is now async processing in between. In case the (stack) caller for instance frees the ref-count wrongly, then this could catch that (when the processing is really async).

**Review by henningandersen — APPROVED** [review]
LGTM.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java
Comment (arteam) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:992]: I believe we already had `refs` declared, so I had to use a different name

**Pull request merged**
